src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/measure/ActionMeter.kt

```diff
@@ -97,42 +109,43 @@ class ActionMeter private constructor(
         observation: (T) -> JsonObject?
     ): T {
         val start = clock.instant()
-        var actionResult: ActionResult = ActionResult.OK
-        var result: T? = null
+        val actionMetricBuilder = ActionMetric.Builder(
+            label = key.label,
+            virtualUser = virtualUser,
+            start = start
+        )
 
         try {
-            result = action()
+            val result = action()
+            result?.let { actionMetricBuilder.observation(observation(result)) }
+            actionMetricBuilder.result(ActionResult.OK)
             return result
         } catch (e: Exception) {
-            actionResult = if (e.representsInterrupt()) {
+            val actionResult = if (e.representsInterrupt()) {
                 ActionResult.INTERRUPTED
             } else {
                 ActionResult.ERROR
             }
+            actionMetricBuilder.result(actionResult)
             throw Exception("Action '${key.label}' $actionResult", e)
         } finally {
-            val metricBuilder = ActionMetric.Builder(
-                label = key.label,
-                result = actionResult,
-                start = start,
-                duration = Duration.between(start, clock.instant())
+            actionMetricBuilder.duration(Duration.between(start, clock.instant()))
+
+            val metricWithoutDrilldown = actionMetricBuilder.build()
+
+            output.write(
+                if (drilldownCondition.test(metricWithoutDrilldown)) {
+                    actionMetricBuilder
```

Comment:
No sure if it is ok to reuse builder tbh, but couldn't think of better solution there WDYT?

Reply:
Reusing the builder is the right pattern. Builders let you "build objects up" in portions by design.
You can avoid `val metricWithoutDrilldown = actionMetricBuilder.build()` entirely. We don't have to rely on the return value from `drilldown`. Builders return `this` purely for syntactic sugar.

- [ ] avoid `metricWithoutDrilldown`

Reply:
Well we rely on metric build by the builder to know if drilldown should be applied or not. The only way I could think of to get rid of `metricWithoutDrilldown` is by using `let` on metric. Pushed the changes right now

Reply:
Aaaah, right, it's the argument for the condition. Makes sense 👍